examples/goroutine_per_partition_consuming_manual_commit/main.go

```diff
@@ -0,0 +1,161 @@
+			time.Sleep(time.Duration(rand.Intn(150)+100) * time.Millisecond)
+			fmt.Printf("Some sort of work done, about to commit t %s p %d\n", topic, partition)
+			err := cl.CommitRecords(context.Background(), recs...)
+			if err != nil {
+				fmt.Printf("Error when committing offsets to kafka err: %v t: %s p: %d offset %d\n", err, topic, partition, recs[len(recs)-1].Offset+1)
+			}
+		}
+	}
+}
+
+type splitConsume struct {
+	mu        sync.Mutex // gaurds assigning / losing vs. polling
+	consumers map[string]map[int32]pconsumer
+}
+
+func (s *splitConsume) assigned(_ context.Context, cl *kgo.Client, assigned map[string][]int32) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for topic, partitions := range assigned {
+		if s.consumers[topic] == nil {
+			s.consumers[topic] = make(map[int32]pconsumer)
+		}
+		for _, partition := range partitions {
+			pc := pconsumer{
+				quit: make(chan struct{}),
+				done: make(chan struct{}),
+				recs: make(chan []*kgo.Record),
+			}
+			s.consumers[topic][partition] = pc
+			go pc.consume(topic, partition, cl)
+		}
+	}
+}
+
+func (s *splitConsume) lost(_ context.Context, cl *kgo.Client, lost map[string][]int32) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	for topic, partitions := range lost {
+		ptopics := s.consumers[topic]
+		for _, partition := range partitions {
+			pc := ptopics[partition]
+			delete(ptopics, partition)
+			if len(ptopics) == 0 {
+				delete(s.consumers, topic)
+			}
+			close(pc.quit)
+			fmt.Printf("Waiting for work to finish t %s p %d\n", topic, partition)
+			<-pc.done
+		}
+	}
+}
+
+func main() {
+	rand.Seed(time.Now().Unix())
+	flag.Parse()
+
+	if len(*group) == 0 {
+		fmt.Println("missing required group")
+		return
+	}
+	if len(*topic) == 0 {
+		fmt.Println("missing required topic")
+		return
+	}
+
+	s := &splitConsume{
+		consumers: make(map[string]map[int32]pconsumer),
+	}
+
+	opts := []kgo.Opt{
+		kgo.SeedBrokers(strings.Split(*brokers, ",")...),
+		kgo.ConsumerGroup(*group),
+		kgo.ConsumeTopics(*topic),
+		kgo.OnPartitionsAssigned(s.assigned),
+		kgo.OnPartitionsRevoked(s.lost),
+		kgo.OnPartitionsLost(s.lost),
+		kgo.DisableAutoCommit(),
+		kgo.BlockRebalanceOnPoll(),
+	}
+
+	cl, err := kgo.NewClient(opts...)
+	if err != nil {
+		panic(err)
+	}
+	// Check connectivity to cluster
+	err = cl.Ping(context.Background())
+	if err != nil {
+		panic(err)
+	}
+
+	s.poll(cl)
+}
+
+func (s *splitConsume) poll(cl *kgo.Client) {
+	for {
+		fetches := cl.PollRecords(context.Background(), 10000)
+		if fetches.IsClientClosed() {
+			return
+		}
+		fetches.EachError(func(_ string, _ int32, err error) {
+			panic(err)
+		})
+		fetches.EachTopic(func(t kgo.FetchTopic) {
+			s.mu.Lock()
+			tconsumers := s.consumers[t.Topic]
+			s.mu.Unlock()
+			if tconsumers == nil {
+				return
+			}
+			t.EachPartition(func(p kgo.FetchPartition) {
+				pc, ok := tconsumers[p.Partition]
+				if !ok {
+					return
+				}
+				select {
+				case pc.recs <- p.Records:
+				case <-pc.quit:
+				}
+			})
+		})
+		s.mu.Lock()
```

Comment:
If I'm reading this correctly, this lock is to ensure onLost is not running?

Reply:
onLost, is having `<-pc.done` that ensures the records are committed before  the goroutine returns. Thus commiting work in progress before rebalance is allowed. Does it makes sense?

Reply:
I'm planning to merge this then add followup commit that adds a few more comments and potentially avoids `EachTopic` (which internally allocates a map). I'm pretty sure on the new API naming.

Reply:
We are unsure if that makes sense our self. Please advice :)

Reply:
I'll merge this now and touch up a few things (file reorganization now that there are two similar examples, potentially avoid FetchTopics, add an example using AutoCommitMarks). Will ping here once done.

Reply:
👍

Reply:
Please check https://github.com/twmb/franz-go/blob/v1.4/examples/goroutine_per_partition_consuming/manual_commit/main.go to see what this example has become :)

I think it's a good bit simpler now to avoid locking.